### PR TITLE
S6931: Support for ASP.NET Core

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
Related to #8870

~~@martin-strecker-sonarsource Please ignore the squash-rebase commit.~~ I rebased.

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Helpers/AspNetMvcHelper.cs [hunk 1]
```diff
@@ -44,18 +44,22 @@ public static class AspNetMvcHelper
         /// </summary>
         public static bool IsControllerMethod(this IMethodSymbol methodSymbol) =>
             methodSymbol.MethodKind == MethodKind.Ordinary
-            && methodSymbol.GetEffectiveAccessibility() == Accessibility.Public
+            && methodSymbol.DeclaredAccessibility == Accessibility.Public
```
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/AspNetMvcHelper.cs:47]: @martin-strecker-sonarsource I changed this condition to cover the UT cases such as
```csharp
    protected class ProtectedController : Controller            // Noncompliant
    {
        [Route("/Index1")] // Secondary
        public IActionResult Index() => View();
    }
```

But I'm not sure anymore. Unless it's a private class I think for internal and protected we should still consider these actions right?

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/AspNetMvcHelper.cs:47]: I would say that controller that are nested classes are not discovered by the MVC routing mechanism, but I haven't tried that. According to the docs

> A controller is an instantiable class, usually [public](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/public), in which at least one of the following conditions is true:

* The class name is suffixed with Controller.
* The class inherits from a class whose name is suffixed with Controller.
* The [Controller] attribute is applied to the class.

> A controller class must not have an associated [NonController] attribute.

[Source](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/actions?view=aspnetcore-8.0#what-is-a-controller)

So I would say, public (and maybe internal) should be enough. If you can make nested controllers work, you can also leave a comment in the tests and here, how this is working.

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/AspNetMvcHelper.cs:47]: Just tested - only actions in public controllers are discoverable.
I'll update the UTs accordingly.

File: analyzers/src/SonarAnalyzer.Common/Rules/RouteTemplateShouldNotStartWithSlashBase.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        {
+            if (!compilationStartContext.Compilation.ReferencesControllers())
+            {
+                return;
+            }
+
+            compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+            {
+                var symbol = (INamedTypeSymbol)symbolStartContext.Symbol;
+                if (symbol.IsControllerType() && symbol.ContainingSymbol is not INamedTypeSymbol)
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/RouteTemplateShouldNotStartWithSlashBase.cs:46]: The `&& symbol.ContainingSymbol is not INamedTypeSymbol` part should go into symbol.IsControllerType()
```suggestion
                if (symbol.IsControllerType())
```

File: analyzers/src/SonarAnalyzer.Common/Rules/RouteTemplateShouldNotStartWithSlashBase.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                }
+            }, SymbolKind.NamedType);
+        });
+
+    private void ReportIssues(SonarSymbolReportingContext context, INamedTypeSymbol controllerSymbol, ConcurrentStack<ControllerActionInfo> actions)
+    {
+        // If one of the following conditions is true, the rule won't raise an issue
+        // 1. The controller does not have any actions defined
+        // 2. At least one action is not annotated with a route attribute or is annotated with a parameterless attribute
+        // 3. At least one route template defined in a route attribute over any action does not start with '/'
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/RouteTemplateShouldNotStartWithSlashBase.cs:68]:
```suggestion
        // 3. There is at least one action with a route template that does not start with '/'
```

File: analyzers/tests/SonarAnalyzer.Test/Rules/RouteTemplateShouldNotStartWithSlashTest.cs [hunk 4]
```diff
@@ -29,8 +29,156 @@ public class RouteTemplateShouldNotStartWithSlashTest
...
+    {
+        var builder = builderCS.AddReferences(AspNetCoreReferences)
+            .AddSnippet($$"""
+                using Microsoft.AspNetCore.Mvc;
+                using Microsoft.AspNetCore.Mvc.Routing;
+
+                public class BasicsController : Controller  // Noncompliant
+                {
+                    {{attribute}} // Secondary
+                    public ActionResult SomeAction() => View();
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/Rules/RouteTemplateShouldNotStartWithSlashTest.cs:105]: These test cases contradict with S6934 "How to fix" option 3).

File: analyzers/tests/SonarAnalyzer.Test/TestCases/RouteTemplateShouldNotStartWithSlash.AspNetCore.cs [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+    [HttpGet("/[action]")]
+    public IActionResult Index3() => View();
+
+    [HttpGet("/[controller]/Index4_1")]
+    [HttpGet("SubPath/Index4_2")] // The relative route
+    public IActionResult Index4() => View();
+}
+
+public class NoncompliantNoControllerRouteController : Controller // Noncompliant {{Change the paths of the actions of this controller to be relative and add a controller route with the common prefix.}}
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/RouteTemplateShouldNotStartWithSlash.AspNetCore.cs:68]: This contradicts suggestion 3) for rule S6934. This is not necessarily a bad thing as option 3) is still not ideal. But if you want to do route value binding for a single action only, it is a viable alternative. You just have to accept the S6931 violation afterward. I think we should cross-link between the two RSpecs.
Please look at the next test case as well. We allow absolute path in this rule as well if there is mixed routing.
/cc @costin-zaharia-sonarsource and @antonioaversa

File: analyzers/tests/SonarAnalyzer.Test/TestCases/RouteTemplateShouldNotStartWithSlash.AspNetCore.cs [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    public IActionResult Index3() => View();
+
+    [HttpGet("/SubPath/Index4_1")]              // Secondary
+    [HttpGet("/[controller]/Index4_2")]         // Secondary
+    public IActionResult Index4() => View();
+}
+
+public class CompliantNoControllerRouteNoActionRouteController : Controller // Compliant
+{
+    public IActionResult Index1() => View(); // Default route -> relative
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/RouteTemplateShouldNotStartWithSlash.AspNetCore.cs:86]:
```suggestion
    public IActionResult Index1() => View(); // Conventional route -> relative
```

File: analyzers/tests/SonarAnalyzer.Test/TestCases/RouteTemplateShouldNotStartWithSlash.AspNetCore.cs [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+    [HttpGet("/SubPath/Index4_1")]
+    [HttpGet("/[controller]/Index4_2")]
+    public IActionResult Index4() => View();
+}
+
+public class CompliantNoControllerRouteEmptyActionRouteController : Controller // Compliant
+{
+    [HttpGet]
+    public IActionResult Index1() => View(); // Empty route -> relative
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/RouteTemplateShouldNotStartWithSlash.AspNetCore.cs:102]:
```suggestion
    public IActionResult Index1() => View(); // Empty route template -> relative conventional routing
```

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM

I left some comments regarding the interaction of S6931 and S6934, but this is nothing that requires action in this PR. It needs to be discussed independently.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/RouteTemplateShouldNotStartWithSlash.AspNetCore.cs (hunk 6 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/RouteTemplateShouldNotStartWithSlash.AspNetCore.cs:86]: fixed here https://github.com/SonarSource/sonar-dotnet/pull/8965/commits/1b3c5531dfa4a2a51793ea3902ab525d469d43b5

File: analyzers/tests/SonarAnalyzer.Test/Rules/RouteTemplateShouldNotStartWithSlashTest.cs (hunk 4 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Rules/RouteTemplateShouldNotStartWithSlashTest.cs:105]: @martin-strecker-sonarsource @antonioaversa   maybe we can exclude in S6931 routes that start with `/[controller]` as they seem pretty intentional.
WDYT?

**Pull request merged**